src/worker/rayon.rs

```diff
@@ -0,0 +1,117 @@
+use core::convert::TryInto;
+use crate::error::Result;
+use crate::idct::dequantize_and_idct_block;
+
+use std::sync::Mutex;
+
+use super::{RowData, Worker};
+use crate::worker::immediate::ImmediateWorker;
+
+pub struct Scoped {
+    inner: Mutex<ImmediateWorker>,
+}
+
+pub fn with_rayon<T>(f: impl FnOnce(&mut dyn Worker) -> T) -> T {
+    rayon::in_place_scope(|_| {
+        let inner = ImmediateWorker::new_immediate();
+        f(&mut Scoped { inner: Mutex::new(inner) })
+    })
+}
```

Comment:
If I'm reading the rayon documentation correctly, this function is equivalent to:
```rust
pub fn with_rayon<T>(f: impl FnOnce(&mut dyn Worker) -> T) -> T {
    let inner = ImmediateWorker::new_immediate();
    f(&mut Scoped { inner: Mutex::new(inner) })
}
```
...which doesn't actually use rayon at all.

Reply:
Specifically this line:
> ...the closure runs on the same thread that calls in_place_scope(). Only work that it spawns runs in the thread pool.

Reply:
Keep in mind this includes work done inside closure `f` (which will be calling `Decoder::decode` etc.). The recursively includes calls to spawn tasks into the thread pool in the `append_row{,s}` impls. The reasoning behind choosing `in_place_scope` specifically:
- We want a scope to guarantee that no more task are running after returning, i.e. cleanup.
- We want `in_place`so as to not require the parameter `f` to have `Sync + 'static` bounds.

Reply:
The problem is that the code ignores the closure argument (a `&rayon::Scope`), which is required to spawn tasks within the scope. See for instance this [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3fe1e5e1fda9660418ed1cb362126f24) which prints the lines out of order